[exporter/tanzuobservability] Add histogram consumer

keep94 opened the pull request:
**Description:** <Describe what has changed.>
Add support for cumulative histograms

**Testing:** <Describe what testing was performed and which tests were added.>
Unit tests.

keep94 committed:
[tanzuobservability exporter] Add histogram consumer.

oppegard commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -263,3 +270,200 @@ func (g *gaugeConsumer) pushSingleNumberDataPoint(
+	var consumer histogramDataPointConsumer
+	switch aggregationTemporality {
+	case pdata.MetricAggregationTemporalityDelta:
+		if h.delta == nil {
+			*errs = append(*errs, errors.New("delta histograms not supported"))
+			return
+		}
```
Replace conditional with `// TODO: follow-on pull request will add support for Delta Histograms`. Fine to propagate error for now until PR is submitted.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -263,3 +270,200 @@ func (g *gaugeConsumer) pushSingleNumberDataPoint(
+		}
+		consumer = h.delta
+	case pdata.MetricAggregationTemporalityCumulative:
+		if h.cumulative == nil {
+			*errs = append(*errs, errors.New("cumulative histograms not supported"))
+			return
+		}
```
Get rid of conditional all together – our metrics exporter should always be able to handle Cumulative Histograms.

keep94 replied:
Done.

oppegard commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -263,3 +270,200 @@ func (g *gaugeConsumer) pushSingleNumberDataPoint(
+	if abort {
+		return
+	}
+	bucketCountLen := len(bucketCounts)
+	for i := 0; i < bucketCountLen; i++ {
```
Inline `len()` to for loop.

keep94 replied:
Done. Used for .. range instead.

oppegard commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,285 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+func verifyCumulativeHistogramDataPointConsumer(t *testing.T, errorOnSend bool) {
+	metric := newMetric("a.metric", pdata.MetricDataTypeHistogram)
+	histogramDataPoint := pdata.NewHistogramDataPoint()
+	histogramDataPoint.SetBucketCounts([]uint64{2, 3, 5, 7})
+	histogramDataPoint.SetExplicitBounds([]float64{2.0, 5.0, 10.0})
```
Can we set the Explicit Bounds first, following by Bucket Counts? I think it has more meaning to know what they bounds are first, before adding datapoints that fall into those buckets. Also, a quick explanation of what this test data means, like:
```
/*
  Create bounds of -Inf to <= 2.0, > 2.0 to <= 5.0, >5.0 to <= 10.0, > 10.0 to +Inf
  Set 2 data points in first bucket, 3 in second, 5 in third, 7 in fourth
*/
```

oppegard commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,285 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+	return result
+}
+
+func verifyCumulativeHistogramDataPointConsumer(t *testing.T, errorOnSend bool) {
```
Can we pull this func into a separate happy-path test, which comes first in the Histogram tests? I'd like the edge cases and error handling tests to come after the happy-path test.

keep94 replied:
Done.

keep94 committed:
Get rid of support for nil histogramDataPointConsumers.

Use for..range where possible.
Sum up bucket counts for wavefront.
Make tests easier to read.

thepeterstone commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -263,3 +270,195 @@ func (g *gaugeConsumer) pushSingleNumberDataPoint(
+	}
+}
+
+func leTagValue(explicitBounds []float64, idx int) string {
```
Would it make sense to rename `idx` to `bucketIndex`?

keep94 replied:
Done.

thepeterstone commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,270 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+func TestHistogramConsumerCumulativeAggregation(t *testing.T) {
+	countAttributeForEachDataPoint := []uint64{2, 5, 10}
+	cumulativeMetric := newHistogramMetricWithDataPoints(
+		"cum.metric",
```
Can we spell out "cumulative.metric" here?

keep94 replied:
Done.

thepeterstone commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,270 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+
+	// Create a histogram metric with missing aggregation attribute
+	metric := newHistogramMetricWithDataPoints(
+		"bad.metric", pdata.MetricAggregationTemporalityUnspecified, nil)
```
Might be nice for future-proofing to name this metric based on _why_ it's bad. `missing.aggregation.metric` or something?

keep94 replied:
Done.

thepeterstone commented on exporter/tanzuobservabilityexporter/metrics_test.go:
```diff
@@ -197,6 +196,270 @@ func TestGaugeConsumerMissingValue(t *testing.T) {
+	metric := newMetric("a.metric", pdata.MetricDataTypeHistogram)
+	histogramDataPoint := pdata.NewHistogramDataPoint()
+
+	// Creates bounds of -Inf to <=2.0; >2.0 to <=5.0; >5.0 to <=10.0; >10.0 to +Inf
```
really like having this comment explicitly spell out the bounds :-)

keep94 replied:
Thank you.

thepeterstone reviewed (approved):
A few minor comments, otherwise LGTM

keep94 commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -263,3 +270,200 @@ func (g *gaugeConsumer) pushSingleNumberDataPoint(
+}
+
+// newHistogramReporting returns a new histogramReporting instance.
+// logger is the logger to use.
```
Done.

keep94 committed:
A few renamings and documentation changes.

bogdandrutu commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -345,3 +350,195 @@ func (s *sumConsumer) pushNumberDataPoint(
+// histogramReporting takes care of logging and internal metrics for histograms
+type histogramReporting struct {
+	logger                   *zap.Logger
+	malformedHistograms      counter
+	leInUse                  counter
+	noAggregationTemporality counter
```
I know it may not feel "natural" for you to use other metrics library, but I would like in the future to change this to use same library as the rest of the collector code :) For the moment please create an issue to track this.

keep94 replied:
Thank you for the feedback. I am a big fan of code reuse rather than reinventing the wheel. Do tell me where to find this library you speak of, so that I may use it instead of my own `counter` type.

keep94 replied:
I created issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6819 for this comment.

bogdandrutu commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -345,3 +350,195 @@ func (s *sumConsumer) pushNumberDataPoint(
+		reporting.LogMalformed(metric)
+		abort = true
+	}
+	if _, ok := tags["le"]; ok {
+		reporting.LogLeTagInUse(metric)
+		abort = true
+	}
```
Usually this should not happen, not sure if you want to be that strict. Up to you.

keep94 replied:
Right. 99.9% of the time there won't be an existing `le` tag.  I'll talk to my tech lead about just overwriting the `le` tag if it already exists.  We'd still like to know if we ever overwrite the `le` tag so I'd like to log/report it when we do.

keep94 replied:
Done. We decided to solve this by moving any pre-existing "le" tag to "_le"

jpkrohling commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -345,3 +350,195 @@ func (s *sumConsumer) pushNumberDataPoint(
+	if options != nil {
+		fixedOptions = *options
+	}
+	fixedOptions.replaceZeroFieldsWithDefaults()
```
You can use this instead:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/51d2a7775b22c07beeb508612564ff9a483cd480/extension/jaegerremotesampling/factory.go#L39-L43

Then you tie it with the factory:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/51d2a7775b22c07beeb508612564ff9a483cd480/extension/jaegerremotesampling/factory.go#L35

When your component is created, this is used as the base for the config, with the options your users provided overriding the default values.

keep94 replied:
What if I got rid of the consumerOptions type entirely?  It only has two fields: `Logger` and `ReportInternalMetrics`.  Any function that expects a `consumerOptions` will take only a zap.Logger. A nil logger means use the nop logger and don't report internal metrics; A non-nil logger means use that logger and report internal metrics.  Note that I would pass nil for logger only in test code. I think this is fine because I wasn't planning on letting the user decide whether or not to report internal metrics in the config file. I was planning on always reporting internal metrics. What do you think?

keep94 replied:
Done. Got rid of consumerOptions type completely.

keep94 commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -345,3 +350,195 @@ func (s *sumConsumer) pushNumberDataPoint(
+
+// newHistogramReporting returns a new histogramReporting instance.
+// log messages are sent to logger.
+func newHistogramReporting(logger *zap.Logger) *histogramReporting {
```
Looking at the code, it seems like I need to implement the MeterImpl interface.  As you know MeterImpl has 3 methods: RecordBatch, NewSyncInstrument, NewAsyncInstrument.  It looks like RecordBatch might do what I need which is to send a bunch of metrics to tanzu observability. But, I am not sure how to implement NewSyncInstrument and NewAsyncInstrument. I don't understand the contract of those two methods.  With that said, I still need a way to count events, so I was looking for something with an `Increment()` or `Inc()` method, but `Measurement` doesn't seem to have this.  In short, I haven't yet figured out how to reuse this code.

jpkrohling replied:
Sorry, my comment was really about the logger parameter. Instead of receiving a logger, you'd receive a `component.TelemetrySettings`, which does contain a logger you can use.

About the reporting types (`histogramReporting`), take a look at my other comment: it points to another component that creates data points for operations performed within the component itself. As Bogdan said, it might be strange to make telemetry data available via OpenCensus/OpenTelemetry when you are at a vendor exporter, but consider that your users might want to monitor the collector in a different way than their workloads.

keep94 replied:
So if I understand correctly, I should use a component.TelemetrySettings instead of a zap.Logger whenever I plan to log events and update internal metrics.  Question:  zap.NewNop() is great for creating a no-op logger when you don't have a *real* logger such as in tests.  Is there a preferred way to create a No-op component.TelemetrySettings?  I could just say `component.TelemetrySettings{Logger: zap.NewNop()}` but if I do that then I am not initializing the TracerProvider and MeterProvider fields.

jpkrohling replied:
Yes: `componenttest.NewNopTelemetrySettings()` from `go.opentelemetry.io/collector/component/componenttest`

https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/componenttest/nop_telemetry.go#L26

keep94 replied:
Done.

keep94 committed:
Fix misspelled word in documentation.

keep94 commented on exporter/tanzuobservabilityexporter/metrics.go:
```diff
@@ -345,3 +350,195 @@ func (s *sumConsumer) pushNumberDataPoint(
+	r.noAggregationTemporality.Inc()
+}
+
+// Report sends the counts in this instancce to wavefront.
```
Done.

keep94 committed:
Get rid of consumerOptions type.

keep94 committed:
Move pre-existing "le" tag to "_le"

jpkrohling commented:
If you want to work on the pending items on a follow-up PR, let me know and I'll merge this as is.

keep94 committed:
Use component.TelemetrySettings instead of logger.

keep94 commented:
I believe I have addressed all pending comments. Would you please have another look? I decided to control whether internal metrics get sent from the top level metricConsumer type rather than from each typedMetricConsumer. Now, typedMetricConsumer.PushInternalMetrics() always sends internal metrics if called. A bool flag controls whether the top level metricConsumer calls PushInternalMetrics() on each typedMetricConsumer.

jpkrohling reviewed (approved):
This looks OK to me. Please, do consider reworking the internal metrics, so that observability for the component itself is done in line with other components.
